_posts/2023-04-21-tensorflow-mnist-classification.adoc

```diff
@@ -0,0 +1,405 @@
+Image bytes can be decoded as a HEX string.
+
+As we will show how to leverage Debezium also for data streaming later in this post, we will initially load into the database only training data set.
+SQL file with training data will be used directly by Postgres container - when it starts, it will load this data into the training table.
+The test data SQL file will be used later on.
+However, the preparation of the data is the same for training as well as test samples and we can prepare both of them in one go.
+
+To prepare the these SQL files, you can use `mnist2sql.py` script from http://TBD[Debezium tensorflow-mnist example]:
+
+[source,bash]
+----
+$ ./mnist2sql.py --download
+----
+
+The script assumes MNIST data sets are available in `postgres` directory.
+When the `--download` parameter is used, script  downloads MNIST data samples into `postgres` directory first.
+The resulting SQL files are also located in the `postgres` directory.
+
+=== Loading streamed data into Tensorflow
+
+The most common Debezium usage is streaming the records to Kafka.
+For loading data from various sources, TensorFlow provides https://www.tensorflow.org/io[TensorFlow I/O] module.
+Besides other sources, it also provides capability to load the data from Kafka.
+There are several ways to do it.
+https://www.tensorflow.org/io/api_docs/python/tfio/IODataset#from_kafka[IODataset.from_kafka()] method loads only existing data from specified Kafka topics.
+For streaming the data there are two classes, currently marked as experimental, https://www.tensorflow.org/io/api_docs/python/tfio/experimental/streaming/KafkaBatchIODataset[KafkaBatchIODataset] and https://www.tensorflow.org/io/api_docs/python/tfio/experimental/streaming/KafkaGroupIODataset[KafkaGroupIODataset].
+Both are very similar and allow to work with streaming data, i.e. they not only read the existing data from Kafka topic, but also wait for new data and eventually pass new records into the TensorFlow.
+If there are no new events within the specified time frame, the streaming finishes.
+
+In all cases obtained records are loaded into TensorFlow as the https://www.tensorflow.org/api_docs/python/tf/data/Dataset[Dataset].
+This is a convenient data structure provided by TensorFlow for building https://www.tensorflow.org/guide/data[data pipelines],  which may include further data transformations or preprocessing.
+
+This sounds great.
+However, the biggest caveat is how the records are represented within the Dataset.
+It seems that these Kafka loaders completely ignore the schema of the records provided by Kafka.
+Keys and values are always represented as the raw bytes.
+This won't be that bad, but even worse, it seems that somewhere on the way the data are converted into the strings (i.e. `toString()` method on the object is called).
+So if you pass e.g. raw image bytes via Kafka, using Kafka `BYTES_SCHEMA`, it would result into something like this:
```

Comment:
```suggestion
However, the most significant caveat is the representation of records within the Dataset.
These Kafka loaders completely ignore the schema of the records provided by Kafka, meaning that keys and values are raw bytes of data.
Additionally, the ingestion pipeline complicates the process by converting these into strings (i.e., `toString()` on the object called).
So if you pass, e.g., raw image bytes via Kafka, using Kafka `BYTES_SCHEMA`, it would result in something like this:
```